File: 03. Stack and Queue/03. Previous Greater Element/main.js
function previousGreaterElement(nums) {
    let stack = [];
    let result = new Array(nums.length).fill(-1);

    for (let i = 0; i < nums.length; i++) {
        while (stack.length > 0 && stack[stack.length - 1] <= nums[i]) {
            stack.pop();
        }
        if (stack.length > 0) {
            result[i] = stack[stack.length - 1];
        }
        stack.push(nums[i]);
    }
    
    return result;
}

// Test
const nums = [4, 5, 2, 10, 8];
console.log("Previous Greater Elements:");
console.log(previousGreaterElement(nums).join(" "));

File: 03. Stack and Queue/03. Previous Greater Element/main.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> previousGreaterElement(vector<int>& nums) {
    stack<int> st;
    vector<int> result(nums.size(), -1);

    for (int i = 0; i < nums.size(); i++) {
        while (!st.empty() && st.top() <= nums[i]) {
            st.pop();
        }
        if (!st.empty()) {
            result[i] = st.top();
        }
        st.push(nums[i]);
    }

    return result;
}

int main() {
    vector<int> nums = {4, 5, 2, 10, 8};
    vector<int> result = previousGreaterElement(nums);

    cout << "Previous Greater Elements:\n";
    for (int num : result) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}
